pdbtools/pdb_addter.py

```diff
@@ -0,0 +1,247 @@
+    elif not is_integer(end):
+        emsg = 'ERROR!! End value must be numerical: \'{}\'\n'
+        sys.stderr.write(emsg.format(end))
+        sys.exit(1)
+    elif not (-999 <= int(end) < 9999):
+        emsg = 'ERROR!! End value must be between -999 and 9998: \'{}\'\n'
+        sys.stderr.write(emsg.format(end))
+        sys.exit(1)
+
+    if step is None:
+        step = 1
+    elif not is_integer(step):
+        emsg = 'ERROR!! Step value must be numerical: \'{}\'\n'
+        sys.stderr.write(emsg.format(step))
+        sys.exit(1)
+    elif int(step) <= 0:
+        emsg = 'ERROR!! Step value must be a positive number: \'{}\'\n'
+        sys.stderr.write(emsg.format(step))
+        sys.exit(1)
+
+    start, end, step = int(start), int(end), int(step)
+
+    if start >= end:
+        emsg = 'ERROR!! Start ({}) cannot be larger than end ({})\n'
+        sys.stderr.write(emsg.format(start, end))
+        sys.exit(1)
+
+    resrange = set(range(start, end + 1))
+    return (fh, resrange, step)
+
+
+def run(fhandle, residue_range, step):
+    """
+    Add TER records within the residue range at frequency identified by step.
+
+    This function is a generator.
+
+    Parameters
+    ----------
+    fhandle : a line-by-line iterator of the original PDB file.
+
+    residue_range : set, list, or tuple
+        The residues describing the range to consider.
+
+    step : int
+        The step at which to insert a TER record.
+
+    Yields
+    ------
+    str (line-by-line)
+        All lines with added TER lines designated by inputs.
+    """
+    prev_res = None
+    res_counter = 0
+    records = ('ATOM')
+    for line in fhandle:
+        if line.startswith(records):
+
+            res_id = line[21:26]  # include chain ID
+            if res_id != prev_res:
+
+                prev_res = res_id
+                res_counter += 1
+                if res_counter - min(residue_range) != 0 and (res_counter - min(residue_range)) % step == 0 and int(line[22:26]) in residue_range:
+                    yield "TER\n"
+
+        yield line
+
+
+add_ter = run
+
+
+def main():
+    # Check Input
+    pdbfh, resrange, step = check_input(sys.argv[1:])
+
+    # Do the job
+    new_pdb = run(pdbfh, resrange, step)
+
+    try:
+        _buffer = []
+        _buffer_size = 5000  # write N lines at a time
+        for lineno, line in enumerate(new_pdb):
+            if not (lineno % _buffer_size):
+                sys.stdout.write(''.join(_buffer))
+                _buffer = []
+            _buffer.append(line)
+
+        sys.stdout.write(''.join(_buffer))
+        sys.stdout.flush()
+    except IOError:
+        # This is here to catch Broken Pipes
+        # for example to use 'head' or 'tail' without
+        # the error message showing up
+        pass
+
+    # last line of the script
+    # We can close it even if it is sys.stdin
+    pdbfh.close()
+    sys.exit(0)
+
+
+if __name__ == '__main__':
+    main()
```

Comment:
I tried to chain `pdb_addter` with `pdb_chainbow` using your example, `test/data/addter_test.pdb` and I have this error:

```
· python pdbtools/pdb_addter.py -1::10 tests/data/addter_test.pdb | python pdbtools/pdb_chainbows.py
Traceback (most recent call last):
  File "pdbtools/pdb_chainbows.py", line 159, in <module>
    main()
  File "pdbtools/pdb_chainbows.py", line 138, in main
    for lineno, line in enumerate(new_pdb):
  File "pdbtools/pdb_chainbows.py", line 106, in run
    chain_map[line[21]] = curchain
IndexError: string index out of range
```

Reply:
I've figured out the reason for this. Even without using `pdb_add_manual_ter` here, the test pdb has a TER line (but not a full TER entry). This is common in pdb files output from Gromacs. These files will look like this at the end:

.
.
ATOM ... 
ATOM ...
TER
ENDMDL

So pdb files output from Gromacs will break `pdb_chainbows` just like `pdb_add_manual_ter` does (before the changes of course I am making of course :smile:).